### PR TITLE
Kube Play - allow setting and overriding published host ports

### PR DESCRIPTION
**Pull request by ygalblum**
#### Does this PR introduce a user-facing change?
Yes

```release-note
Kube Play - Add --publish flag to set or override port publishing
```

### CODE REVIEW

**Comment by ygalblum** [comment]
This PR is the first step in addressing #16595

File: test/e2e/play_kube_test.go
Comment (vrothberg) [inline test/e2e/play_kube_test.go:4556]:
```suggestion
		curlTest := podmanTest.Podman([]string{"run", "--network", "host", NGINX, "curl", "-s", "localhost:19000"})
```

For other tests as well.  We have consts for the images such that we can quickly address flakes, do updates, etc.

**Review by vrothberg — COMMENTED** [review]
LGTM
@Luap99 @umohnani8 PTAL

File: test/e2e/play_kube_test.go
Comment (ygalblum) [reply test/e2e/play_kube_test.go]: Funny thing is that I found this container image through the const, but forgot to use it. Fixed

File: test/e2e/play_kube_test.go
Comment (Luap99) [inline test/e2e/play_kube_test.go:4586]: this is a bug and should be fixed? assuming this is already the case before this PR could you fill a issue for it

File: test/e2e/play_kube_test.go
Comment (Luap99) [inline test/e2e/play_kube_test.go:4558]: exit codes are positive

File: test/e2e/play_kube_test.go
Comment (Luap99) [inline test/e2e/play_kube_test.go:4582]: same here

File: test/e2e/play_kube_test.go
Comment (Luap99) [inline test/e2e/play_kube_test.go:4576]: this test may conflict with https://github.com/containers/podman/pull/16766

File: test/e2e/play_kube_test.go
Comment (Luap99) [inline test/e2e/play_kube_test.go:4572]: given that we want to have the tests fast as poosible I think it would be better to not run a new container, just connect in the go code to this address, same for the other tests

**Review by Luap99 — CHANGES_REQUESTED** [review]

File: cmd/podman/kube/play.go
Comment (ygalblum) [reply cmd/podman/kube/play.go:155]: Done

File: docs/source/markdown/podman-kube-play.1.md.in
Comment (ygalblum) [reply docs/source/markdown/podman-kube-play.1.md.in]: Done

File: pkg/domain/infra/abi/play.go
Comment (ygalblum) [reply pkg/domain/infra/abi/play.go]: I've changed the code so that it starts with the list of published ports specified in the command line and adds to the list ports from the YAML file that are not listed in the command line. The matching is still done according to the `ContainerPort`.

File: test/e2e/play_kube_test.go
Comment (ygalblum) [reply test/e2e/play_kube_test.go]: I've changed the code to not use Podman for curl so, this comment is no longer relevant

File: test/e2e/play_kube_test.go
Comment (ygalblum) [reply test/e2e/play_kube_test.go]: Fixed

File: test/e2e/play_kube_test.go
Comment (ygalblum) [reply test/e2e/play_kube_test.go]: Done

**Comment by ygalblum** [comment]
I've fixed the code according to the comments. But, I think we need to align the functionality between the two changes this and #16766 before we proceed with either of them

**Comment by TomSweeneyRedHat** [comment]
A number of red tests.  The changes LGTM in general, I'll let others decide on the thumb-wrestling between this and #16766

**Comment by ygalblum** [comment]
@Luap99 I've fixed the code according to your comment. Please see my explanation here: https://github.com/containers/podman/pull/16880#discussion_r1056938410.
@dcermak wrote in #16595 that he thinks this PR should go in first. So, if you agree, let's continue here so he can complete his change as well

File: pkg/domain/infra/abi/play.go
Comment (Luap99) [inline pkg/domain/infra/abi/play.go:1028]: this works better but ignores the protocol, if you have a udp port in the yaml and also forward tcp via cli it must use both ports
if you fix this please also add a test for this case

File: test/e2e/play_kube_test.go
Comment (ygalblum) [reply test/e2e/play_kube_test.go:4665]: Discussed. This PR should be merged first

File: pkg/domain/infra/abi/play.go
Comment (ygalblum) [reply pkg/domain/infra/abi/play.go]: Fixed and added two tests at the bottom

File: test/e2e/play_kube_test.go
Comment (Luap99) [inline test/e2e/play_kube_test.go:4872]: this needs an inspect or something to make sure that both ports are defined, your test only check for the tcp port below

File: test/e2e/play_kube_test.go
Comment (Luap99) [inline test/e2e/play_kube_test.go:4883]: same here

File: pkg/domain/infra/abi/play.go
Comment (Luap99) [inline pkg/domain/infra/abi/play.go:1025]: necessary newline

**Review by Luap99 — COMMENTED** [review]
mostly looks good, the test should make sure both ports are present

File: test/e2e/play_kube_test.go
Comment (ygalblum) [reply test/e2e/play_kube_test.go:4918]: I changed the test to test both UDP and TCP connections. So, I guess there's no need for inspect, right?

File: test/e2e/play_kube_test.go
Comment (ygalblum) [reply test/e2e/play_kube_test.go:4930]: I changed the test to test both UDP and TCP connections. So, I guess there's no need for inspect, right?

File: pkg/domain/infra/abi/play.go
Comment (ygalblum) [reply pkg/domain/infra/abi/play.go]: Right, fixed

**Comment by rhatdan** [comment]
/approve
@ygalblum Looks like your test is failing?

**Comment by ygalblum** [comment]
The tests pass locally. Looking at the failure logs, it seems that the containers are started successfully, but then the test code fails to connect to the port. In addition, other flavors of the suite pass. So, it might be a difference in the setup that causes the failures. @containers/podman-maintainers Can you think of a reason why they fail in the CI?

**Comment by Luap99** [comment]
CI systems are always much slow compared to local. You need to add some retry mechanism for the connect in the test. If we do a `podman run -d` or in this case `podman play kube` it will return as soon as the container process is started but that does not mean that this container process is ready to accept incoming connections.
something like this https://github.com/containers/podman/blob/98d95f0735cd68a678f56f9e066a838e7f3aa86d/test/e2e/network_test.go#L514-L526

**Comment by ygalblum** [comment]
@Luap99 Thanks for the tip. I've updated the test code

**Comment by rhatdan** [comment]
Lint is not happy with this change.

**Comment by ygalblum** [comment]
> Lint is not happy with this change.

Thanks, fixed

**Comment by ygalblum** [comment]
@containers/podman-maintainers can someone please restart the failed test. This time it is unrelated to my changes

**Comment by mheon** [comment]
Done

File: test/e2e/play_kube_test.go
Comment (Luap99) [reply test/e2e/play_kube_test.go:4965]: sure that works too

**Review by Luap99 — COMMENTED** [review]
LGTM

**Comment by ygalblum** [comment]
@Luap99 I've made all the change and all tests are done and passing, but the PR still has the "Changes requested" state. Can you please re-review?

**Review by Luap99 — APPROVED** [review]
LGTM

**Comment by vrothberg** [comment]
This is already in the v4.4 branch

**Pull request merged**
